enhancements/update/dont-require-registry-during-boot-and-upgrade.md

```diff
@@ -0,0 +1,201 @@
+---
+title: dont-require-registry-during-boot-and-upgrade
+authors:
+- "@jhernand"
+reviewers:
+- "@avishayt"
+- "@danielerez"
+- "@mrunalp"
+- "@nmagnezi"
+- "@oourfali"
+approvers:
+- "@sdodson"
+- "@zaneb"
+- "@LalatenduMohanty"
+api-approvers:
+- "@sdodson"
+- "@zaneb"
+- "@deads2k"
+- "@JoelSpeed"
+creation-date: 2023-09-21
+last-updated: 2023-09-21
+tracking-link:
+- https://issues.redhat.com/browse/RFE-4482
+see-also:
+- https://issues.redhat.com/browse/OCPBUGS-13219
+- https://github.com/openshift/enhancements/pull/1481
+- https://github.com/openshift/cluster-network-operator/pull/1803
+replaces: []
+superseded-by: []
+---
+
+# Don't require registry during boot and upgrade
+
+## Summary
+
+Ensure that clusters don't require a registry server to boot or upgrade when
+all the required images have already been pulled.
+
+## Motivation
+
+Currently during reboots and upgrades clusters may need to contact the image
+registry servers, even if the images have already been pulled. This complicates
+things for clusters that are completely disconnected or that have an slow or
+unreliable connection to the image registry servers.
+
+### User Stories
+
+#### Boot without registry
+
+As the administrator of a cluster that has all the required images already
+pulled in all the nodes, I want to be able to reboot it without requiring
+access to a registry server.
+
+### Upgrade without registry
+
+As the administrator of a cluster that has all the required images already
+pulled in all the nodes, I want to be able to upgrade it without requiring
+access to a registry server.
+
+### Goals
+
+Ensure that clusters don't require a registry server to boot or upgrade when
+all the required images have already been pulled.
+
+### Non-Goals
+
+It is not the goal of this enhancement to add a mechanism to ensure that
+required images are available. That is the subject of the [pin and pre-load
+images](https://github.com/openshift/enhancements/pull/1481) enhancement.
```

Comment:
i think it's still responsible for confirming the requested images have been pulled+pinned (by checking the status information described in the other EP) but the actual mechanics of pulling+pinning i agree are not part of this EP.

Reply:
Checking if the image has been pulled or not is something that each OCP component should do completely independent of whatever #1481 says. For example, for the OVN upgrade it is a matter of not using the `Always` pull policy (it used to do, was fixed in https://github.com/openshift/cluster-network-operator/pull/1803). For MCO it is a matter of not fetching data directly from the registry. I don't think any of these components should directly use the API introduced in #1481.

Reply:
> Checking if the image has been pulled or not is something that each OCP component should do completely independent of whatever https://github.com/openshift/enhancements/pull/1481 says.

There are two categories of image usage we need to worry about:

1) standard CRIO-based usage (component defines a pod, relies on crio to pull the image for the pod).  I would not expect every component that is creating pods to have to check that its images are pulled/pinned.  I would expect this new upgrade logic to ensure all the images for the new payload are pulled+pinned before allowing the CVO to update the component manifests to reference the new images.  And as you say, those components not using `pull always` is a piece of this.

2) components that directly talk to registries - pulling + pinning has no bearing on them and any component that does this is going to have to be addressed explicitly.  This means things like the system imagestream that defines the `cli`, `test`, etc imagestream tags.  The imagestream controller imports those tags by talking directly to the external registry that hosts the release payload.  Those imports will fail if no registry is available to do the import from.  That won't necessarily block a bootstrap or upgrade, but it's something that needs to be taken into account, even if that means we have to say "yeah the system imagestreamtags will fail to import if you don't have a registry and that means commands like "oc debug" will fail unless you explicitly provide the image to use" (but hopefully we can do better.  Again there's some muddiness between when is the goal "run with no registry" and when is the goal "we have a registry but we want to pre-pull the content from that registry".  The latter is far easier to reason about(and less risky) than the former.

I don't know what interaction MCO has w/ the registry so i can't speculate on the impact or mitigation there.

Reply:
I think we are talking about the same things.

I think that your first point is the same that I say about not using the `Always` pull policy. We don't really need to change anything as CRI-O already knows to not contact the registry if the image is already pulled. Except if the pull policy is `Always`, in that case it will contact the registry anyhow. All we need here is make sure that no pod needed for the upgrade to succeed uses the `Always` pull policy.

I used MCO as an example of your second point: it fetches directly from the registry server parts of the release image to check the kind of upgrade. I used to have a fallback, but in 4.13 that fallback was removed. I believe that @cgwalters can explain that better.

I have no objection to enumerate here what should be done for other cases, like image streams. But am not very familiar with image streams in particular. Would be nice if you can elaborate on what would be needed to have that working when there is no registry server available.

Reply:
> Would be nice if you can elaborate on what would be needed to have that working when there is no registry server available.

The short answer is at least as it is implemented today, it just can't work.  You can't import an image tag into an imagestream if you can't reach the registry that's hosting the image.

So it becomes a question of what are the implications if any of these imagestreamtags aren't imported successfully:
https://github.com/openshift/cluster-samples-operator/blob/master/manifests/08-openshift-imagestreams.yaml

And the same for any other imagestreams that are being created on the cluster.

Reply:
The cluster-samples-operator has been problematic in the past because it effectively ties "cluster install success" with "access to upstream registry", turning flakes in that into install flakes.

Is there actually no configuration for that operator today?  Is it supported to edit the imagestream objects directly and turn off the `scheduled` flag and bring the pulls under admin control?

Reply:
the static imagestreams i'm talking about in this case are not actually managed by the samples operator, they are just imagestream manifest yamls that happen to live in the samples operator bundle and are applied by the CVO.

I do not believe their import success(or failure) impacts the cluster install result, though i'm not positive.

For the sample imagestreams the operator *does* manage directly(is responsible for creating and may care about the result of), i thought we stopped making the samples operator fail the install when those imagestreams failed to import, but i don't remember the current state for sure.  It's owned by davd peraza these days so he'd be the one to consult on implications there.

Reply:
Added a paragraph about this in the open questions section.